torch/csrc/jit/codegen/cuda/lower_predicate.cpp

```diff
@@ -48,9 +48,54 @@ class ConditionalFromPredicateModifier : public kir::IrVisitor {
       setWritePredicate(expr, conditional);
     }
 
+    // Note: [Predicate Inversion for CpAsync]
+    // Today for vectorized support the pattern is:
+    // Initialize buffer -> predicated load
+    // For memcpy async:
+    //    If we initialized and then loaded (without sync) it would be undefined
+    //    behavior.
+    // Initialize only the "virtual out of boundary" accesses.
+    //  Memory allocated, but outside the virtual tensor space.
+    //  Virtual tensor space today is effectively what would be allocated in
+    //  global memory. Then only copy the "within bound" accesses.
+    // This is a WAR today based on how our system is set up.
+    //    We would want to have a separate concept of SMEM space from Virtual or
+    //    GMEM space, so that we know we're only working with the allocated
+    //    SMEM.
+    //  If we hit outside the allocated SMEM bad things happen.
+    // Today asserting in predicate removal making sure that the virtual and
+    // SMEM boundaries line up based on the IterDomains.
+    //
+    // TODO: in a follow up we need to extend the predicate
+    //  infrastructure to generate predicate for both gmem
+    //  and smem, and the predicate removal will need to
+    //  be extended as well for the perf critical regions.
+    if (isPredicatedInitForCpAsync(expr)) {
+      invertPredicateForGmemToSharedMemInitialize(expr);
```

Comment:
Just copying comments from #1440 

```
naoyam 23 days ago

Not following why this is done. Can you have some illustrative example?

Owner
 csarofeen 22 days ago

https://github.com/csarofeen/pytorch/pull/1440/files#r844288107
Has some details, but agree would be great to have an example of the code we want to generate.
```

Reply:
Sure. This mainly concerns the way we currently do the initialization - predicate combination. The code gen today would produce:
```cpp
  __shared__ T0;
  T0 = 0;  // expr0
  if(pred)
    T0 = ...  // expr1
```
In the case where expr1 is cp.async. The code behavior is officially not defined. So the current WAR solution is to generate:
```cpp
  __shared__ T0;
  if(!pred)
    T0 = 0;  // expr0
  if(pred)
    T0 = ...  // expr1
```
To avoid race condition. 

This is still a WAR but should be safe from cp.async related race condition. In follow ups will build up the correct initialize sequence gradually.

Reply:
Yes we can take predicate as an input. 

Just wanted to discuss a few more details before we proceed to formalizing. 

Taking predicate this way has many other benefits in general, the exact profitability depends on how well we do other optimizations, which I think we should definitely explore but I guess probably separately in another PR. 

Combining the initialization and the ldst into one resource string call may not always be what we want to do though. It's cleaner than the way it is done in this PR at the moment but in the case of cp.async we'd want to be tear them apart again and optimize further at some point:

This will be a good further optimization when we are relatively deeper into compute bound regime (Looks like won't be a limitation until we are >70%). Both the init expr and the cp.async expr will write into shared mem so we'd not want to put them back to back but we'd want to manually squeeze other ops in between these two to avoid throttling the shared memory. 

It's probably either way today but just wanted to mention that keeping the option of separating init and cp.async into two expressions in our IR would be helpful later on as well.